recon/src/protocol.rs

```diff
@@ -735,10 +783,35 @@ where
     }
 
     async fn persist_all(&mut self) -> Result<()> {
+        if self.event_q.is_empty() {
+            return Ok(());
+        }
+
         let evs: Vec<_> = self.event_q.drain(..).collect();
-        if !evs.is_empty() {
-            self.recon.insert(evs).await?;
+
+        let mut batch = self.recon.insert(evs).await.context("persisting all")?;
+        if !batch.invalid.is_empty() {
+            if let Ok(cnt) = batch.invalid.len().try_into() {
+                self.recon.metrics().record(&InvalidEvents(cnt))
+            }
+            tracing::warn!(
+                "Recon discovered data it will never allow. Hanging up on peer {}",
+                self.config.peer_id
+            );
+            bail!("Received unknown data from peer: {}", self.config.peer_id);
         }
+
+        if !batch.pending.is_empty() {
+            let (added, remaining_capacity) = self.pending.track_pending(&mut batch.pending);
```

Comment:
Details depend on how we change the pending cache, but we should probably track every item we re-add as well